Set proxy port in policy map

rlenglet opened the pull request:
Set the proxy port in policy map entries.
Reorganize `regenerateBPF` to keep ensuring that proxy redirects are always realized when redirecting traffic to them.

Output proxy port in `cilium bpf policy get` CLI, e.g.:
```
$ sudo cilium bpf policy get 29094
DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS   
Egress      container:id.client           ANY          NONE         0       0         
Egress      reserved:init                 ANY          NONE         0       0         
Egress      container:id.server           ANY          NONE         0       0         
            container:id.service1                                                          
Egress      reserved:world                ANY          NONE         0       0         
Ingress     reserved:host                 80/TCP       11555        0       0         
Ingress     container:id.client           80/TCP       11555        0       0         
Egress      reserved:host                 ANY          NONE         0       0         
Ingress     reserved:host                 ANY          NONE         0       0         
Egress      reserved:cluster              ANY          NONE         0       0         
Egress      reserved:health               ANY          NONE         0       0         
```

Remove the generation and lookup of the `CFG_L3L4_INGRESS` / `CFG_L3L4_EGRESS` arrays in BPF.
Only define `IPCACHE4_PREFIXES` / `IPCACHE6_PREFIXES` macros in BPF if LPM maps are not supported.
Therefore, on Linux kernels versions where LPM maps are supported (≥4.11), a policy regeneration shouldn't require a BPF regeneration anymore.

Fixes: https://github.com/cilium/cilium/issues/4711
Signed-off-by: Romain Lenglet <romain@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4724)
<!-- Reviewable:end -->

rlenglet committed:
bpf: Remove CFG_L3L4_INGRESS/CFG_L3L4_EGRESS generation and lookup

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
bpf: Set the proxy port in policy map entries

Fixes: https://github.com/cilium/cilium/issues/4711
Signed-off-by: Romain Lenglet <romain@covalent.io>

ianvernon commented on cilium/cmd/helpers.go:
```diff
@@ -269,7 +269,9 @@ func updatePolicyKey(cmd *cobra.Command, args []string, add bool) {
 		entry := fmt.Sprintf("%d %d/%s", label, port, u8p.String())
 		if add == true {
-			if err := policyMap.Allow(label, port, u8p, parsedTd); err != nil {
+			// TODO: Support passing a non-zero proxy port.
```
This is still a TODO, but should we ever do this? Wouldn't this require bootstrapping the proxy and having the proxy configured to accept traffic on that port?

rlenglet replied:
Removed.

tgraf reviewed (approved):
This is a great change as it requires less recompilation as well!

Looks great overall. In case you need to respin, consider using a struct for the value of `desiredMapState` for readability. It's not clear that the map value is the redirect port at first glance.

rlenglet committed:
bpf: Only define IPCACHE4/6_PREFIXES macros if LPM is not supported

Signed-off-by: Romain Lenglet <romain@covalent.io>

joestringer reviewed (approved):
This is awesome. Only thing that made me nervous was the shifting of the `ProxyWaitGroup` / `syncPolicyMap` logic in `pkg/endpoint/bpf.go`.

Probably `bpf/lxc_config.h` needs an update with this change, too.

rlenglet committed:
cli: Output proxy port in "cilium bpf policy get"

Signed-off-by: Romain Lenglet <romain@covalent.io>

rlenglet committed:
tests: Remove tracing test cases from KafkaPolicies

Because of the random scheduling of the pods under test, the
tracing commands may run on a different node from all the pods,
where Cilium may be out of date re: imported policies and
identities, resulting in tracing failures.
This made those tests unreliable.

Signed-off-by: Romain Lenglet <romain@covalent.io>
